google/cloud/aiplatform/utils/worker_spec_utils.py

```diff
@@ -152,17 +161,52 @@ def pool_specs(
         spec_order = [
             self.chief_spec,
             self.worker_spec,
-            self.parameter_server_spec,
+            self.server_spec,
             self.evaluator_spec,
         ]
         specs = [s.spec_dict for s in spec_order]
+        last_non_empty_seen = False
         for i in reversed(range(len(spec_order))):
             if spec_order[i].is_empty:
-                specs.pop()
+                if not last_non_empty_seen:
+                    specs.pop()
+                else:
+                    specs[i] = {}
             else:
-                break
+                last_non_empty_seen = True
         return specs
 
+    def reduction_server_pool(
+        self,
+        reduction_server_replica_count: int = 0,
+        reduction_server_machine_type: str = "n1-highcpu-16",
+    ):
+        """Parameterizes Config to support only reduction server replicas.
+
+        Args:
+            reduction_server_replica_count (int):
+                The number of reduction server replicas, default is 0.
+            reduction_server_machine_type (str):
+                The type of machine to use for reduction server, default is `n1-highcpu-16`.
+
+        Raise:
+            ValueError if chief worker is not set
+        """
+        reduction_server_spec = _WorkerPoolSpec(
+            replica_count=reduction_server_replica_count,
+            machine_type=reduction_server_machine_type,
+            container_uri=REDUCTION_SERVER_CONTAINER_URI,
+        )
+
+        if self.chief_spec.is_empty:
+            raise ValueError("reduction server can not be set without chief worker.")
+
+        return _DistributedTrainingSpec(
```

Comment:
Hm, you are returning a new _DistributedTrainingSpec, but it will hold pointers to the existing `chief_spec`, `chief_spec` since Python passes objects by reference. This seems potentially dangerous and can cause hard to debug situations down the road.

I would say either mutate the existing instance or do a deep copy to create a new instance of `_DistributedTrainingSpec` with no shared pointers.

@sasha-gitg may know a canonical Python way to handle this.